// node-rx/dist/node-rx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var store_1 = require("./store");
var effects_1 = require("./effects");
var NodeRx = (function () {
    function NodeRx(initialReducer, initialState, effectsInstance) {
        var INITIAL_REDUCER = NodeRx.initialReducerFactory(initialReducer);
        var INITIAL_STATE = NodeRx.initialStateFactory(initialState, INITIAL_REDUCER);
        this.dispatcher = new store_1.Dispatcher();
        this.reducer = new store_1.Reducer(this.dispatcher, INITIAL_REDUCER);
        this.state = new store_1.State(INITIAL_STATE, this.dispatcher, this.reducer);
        this.store = NodeRx.storeFactory(this.dispatcher, this.reducer, this.state);
        // effects
        this.actions = new effects_1.Actions(this.dispatcher);
        this.effectSubscription = new effects_1.EffectsSubscription(this.store, null, effectsInstance);
    }
    NodeRx.initialReducerFactory = function (reducer) {
        if (typeof reducer === 'function') {
            return reducer;
        }
        return store_1.combineReducers(reducer);
    };
    NodeRx.initialStateFactory = function (initialState, reducer) {
        if (!initialState) {
            return reducer(undefined, { type: store_1.Dispatcher.INIT });
        }
        return initialState;
    };
    NodeRx.storeFactory = function (dispatcher, reducer, state$) {
        return new store_1.Store(dispatcher, reducer, state$);
    };
    return NodeRx;
}());
exports.NodeRx = NodeRx;
